### PR TITLE
fix(file source): Fix tailing problem when source number greater than 512

### PR DESCRIPTION
**Pull request by honganan**
fix: #17360

### CODE REVIEW

File: lib/file-source/src/file_server.rs [hunk 1]
```diff
@@ -305,17 +305,19 @@ where
...
 
-            let start = time::Instant::now();
-            let to_send = std::mem::take(&mut lines);
-            let result = self.handle.block_on(chans.send(to_send));
-            match result {
-                Ok(()) => {}
-                Err(error) => {
-                    error!(message = "Output channel closed.", %error);
-                    return Err(error);
+            if !lines.is_empty() {
```
[the remark this answers is not in the document]
Comment (honganan) [reply lib/file-source/src/file_server.rs:308]: Ah no, it's not needed in this issue. Just a positive change making during my test, like you said.

We may remove it to narrow down changes, how was your suggestion?

**Comment by jszwedko** [comment]
Thanks @honganan ! I'll make sure this gets more reviews next week.

File: lib/file-source/src/file_server.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jszwedko) [reply lib/file-source/src/file_server.rs:308]: Yeah, could we split out this change into a separate PR since it is unrelated?

**Review by jszwedko — COMMENTED** [review]
Thanks for this change @honganan ! We discussed this a bit and think it is a reasonable workaround for now. Longer term, we have some desire to re-architect the `file` source so that it doesn't spawn additional threads, at which point this higher blocking thread limit shouldn't be needed.

Can you split out the other change from this blocking thread limit change? Otherwise this looks good to merge in.

**Review by lukesteensen — APPROVED** [review]
This seems like a reasonable workaround for now. Longer term, it'd be nice to look into:

1. Is there anything we can do to reduce the need to run so many file sources? Maybe a feature like applying tags based on which glob pattern was matched to make routing easier.
2. Reworking the file source such that it does not require a long-running blocking thread like this (we've talked about this for a long time).

File: lib/file-source/src/file_server.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (honganan) [reply lib/file-source/src/file_server.rs]: Done.

**Comment by honganan** [comment]
> This seems like a reasonable workaround for now. Longer term, it'd be nice to look into:
> 
> 1. Is there anything we can do to reduce the need to run so many file sources? Maybe a feature like applying tags based on which glob pattern was matched to make routing easier.
> 2. Reworking the file source such that it does not require a long-running blocking thread like this (we've talked about this for a long time).

I have thought about same like these two solution.
1. For us, we have many `source` belong to a batch of service, every service need to send logs to different Kafka topic based on log file which defined by user (e.g. warn.log and error.log to topic A,  info.log and some others to topic B).

If different user can define their own `source` which carry additional topic info in self defined `source.metadata`, and then applying mechanism like your idea to merge sources, it would work, but seems not graceful enough. We may need to consider more about how to do.

2. That sounds like exciting.

**Review by jszwedko — APPROVED** [review]
Thanks @honganan !

**Pull request merged**